gallery_dl/extractor/nana.py

```diff
@@ -0,0 +1,115 @@
+# -*- coding: utf-8 -*-
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 2 as
+# published by the Free Software Foundation.
+
+"""Extractors for https://nana.my.id/"""
+
+from .common import GalleryExtractor, Extractor, Message
+from .. import text, exception
+import json
+
+
+class NanaGalleryExtractor(GalleryExtractor):
+    """Extractor for image galleries from nana.my.id"""
+    category = "nana"
+    directory_fmt = ("{category}", "{title}")
+    pattern = r"(?:https?://)?nana\.my\.id/reader/([^/?#]+)"
+    test = (
+        (("https://nana.my.id/reader/"
+          "059f7de55a4297413bfbd432ce7d6e724dd42bae"), {
+            "pattern": r"https://nana\.my\.id/reader/"
+                       r"\w+/image/page\?path=.*\.\w+",
+            "title"  : "Everybody Loves Shion",
+            "artist" : "fuzui",
+            "tags"   : list,
+            "count"  : 29,
+        }),
+        (("https://nana.my.id/reader/"
+          "77c8712b67013e427923573379f5bafcc0c72e46"), {
+            "pattern": r"https://nana\.my\.id/reader/"
+                       r"\w+/image/page\?path=.*\.\w+",
+            "title"  : "Lovey-Dovey With an Otaku-Friendly Gyaru",
+            "artist" : "Sueyuu",
+            "tags"   : ["Sueyuu"],
+            "count"  : 58,
+        }),
+    )
+
+    def __init__(self, match):
+        self.gallery_id = match.group(1)
+        url = "https://nana.my.id/reader/" + self.gallery_id
+        GalleryExtractor.__init__(self, match, url)
+
+    def metadata(self, page):
+        title = text.unescape(
+            text.extract(page, '</a>&nbsp; ', '</div>')[0])
+        artist = text.unescape(text.extract(
+            page, '<title>', '</title>')[0])[len(title):-10]
```

Comment:
If we could assume that every gallery has an artist and each `<title>` does always contain `TITLE by ARTIST - Nana ナナ`, this could be simplified a bit

``` python
title, _, artist = text.unescape(text.extract(
            page, '<title>', '</title>')[0]).rpartition(" by ")
```

Otherwise unescaping `title` first and then using its `len()` will produce wrong results if `title` contains any escaped characters, so move the unescape to line 54.

Reply:
is this okay?

``` python
title, _, artist = text.unescape(text.extract(
            page, '<title>', '</title>')[0])[:-10].rpartition(" by ")
```
but if <title> doesn't contain " by artist", this will make `title` empty and `artist` the `title`

Reply:
> but if <title> doesn't contain " by artist", this will make `title` empty and `artist` the `title`

Yeah, that's the issue. And it gets even worse when there is no " by artist" *and* the title contains ` by `.

I thought maybe you knew how this site behaves or had found a gallery without " by artist", so I wanted to at least bring it up.


Anyway, lets keep it the way you had it but move `text.unescape()` down to the return statement.

Reply:
> move `text.unescape()` down to the return statement.

something like this? then i guess i would also need to `unescape` the artist in the return statement. should i use `replace` instead of `len(title)` ?
[link](https://nana.my.id/reader/9e2ad2d84e0fc19d627f68d05cdfe9ee30a3e15a) with escape character

``` python
        title = text.extract(page, '</a>&nbsp; ', '</div>')[0]
        artist = text.extract(page, '<title>', '</title>')[0][len(title):-10]
        tags = text.extract(page, 'Reader.tags = "', '"')[0]

        return {
            "gallery_id": self.gallery_id,
            "title"     : text.unescape(title),
            "artist"    : text.unescape(
                artist[4:]) if artist.startswith(" by ") else "",
            "tags"      : tags.split(", ") if tags else (),
            "lang"      : "en",
            "language"  : "English",
        }
```

> found a gallery without " by artist"

I couldn't find any


Tachiyomi does this
``` java
        author = img.attr("alt")
            .replace("$title by ", "")
            .ifBlank { null }
```

Reply:
> something like this?

Yeah, that's what I meant, but I also realized that this is not actually necessary since your original code is unescaping the title twice (for both `title` and `artist`) so there won't be any issues with `len()` differences. (*)

Just keep your code as it is and I'm going to merge it. It's doing a bit more than necessary but whatever.

> should i use replace instead of len(title) ?

No, I'd rather slice than replace.

(*)
I thought there would be an issue with a title like `foo&#039;s bar` since you are using `len("foo's bar")` to slice into `foo&#039;s bar by artist  - Nana ナナ`, but the second `foo&#039;s bar` would already be unescaped by then as well.

Reply:
> I thought there would be an issue with a title like `foo&#039;s bar` since you are using `len("foo's bar")` to slice into `foo&#039;s bar by artist - Nana ナナ`, but the second `foo&#039;s bar` would already be unescaped by then as well.

I was gonna bring this up, but I thought I was being wrong so I didn't.. anyways thanks

I will update the change to line 55